src/bgw/scheduler.c

```diff
@@ -0,0 +1,402 @@
+/*
+ * This is a scheduler that takes background jobs and schedules them appropriately
+ *
+ * Limitations: For now the jobs are only loaded when the scheduler starts and are not
+ * updated if the jobs table changes
+ *
+ * TODO: right now jobs are not prioritized when slots available. Should prioritize in
+ * asc next_start order.
+ *
+*/
+#include <postgres.h>
+
+#include <miscadmin.h>
+#include <postmaster/bgworker.h>
+#include <storage/ipc.h>
+#include <storage/latch.h>
+#include <storage/lwlock.h>
+#include <storage/proc.h>
+#include <storage/shmem.h>
+#include <utils/jsonb.h>
+#include <utils/timestamp.h>
+#include <utils/snapmgr.h>
+#include <utils/memutils.h>
+#include <access/xact.h>
+#include <pgstat.h>
+
+#include "extension.h"
+#include "scheduler.h"
+#include "job.h"
+#include "job_stat.h"
+#include "version.h"
+#include "compat.h"
+
+#define MAX_TIMEOUT (5*1000L)
+#define SCHEDULER_APPNAME "TimescaleDB Background Worker Scheduler"
+
+TS_FUNCTION_INFO_V1(ts_bgw_scheduler_main);
+
+
+/* has to be global to shutdowns jobs on exit */
+static List *scheduled_jobs = NIL;
+
+typedef struct ScheduledBgwJob
+{
+	BgwJob	   *job;
+	TimestampTz next_start;
+	TimestampTz timeout_at;
+
+	bool		enabled;
+	BackgroundWorkerHandle *handle;
+} ScheduledBgwJob;
+
+static void scheduled_bgw_job_mark_stopped(ScheduledBgwJob *job);
+
+static inline void
+bgw_scheduler_on_postmaster_death(void)
+{
+	/* don't call exit hooks cause we want to bail out quickly */
```

Comment:
We need a comment on how: in the case of postmaster death, we no longer care about the state of shared memory, it'll be invalid.

Reply:
fixed